calico_upgrade/Makefile

```diff
@@ -0,0 +1,326 @@
+# Version directory
+CALICO_UPGR_DIR=$(dir $(realpath $(lastword $(MAKEFILE_LIST))))
+VERSIONS_FILE?=$(CALICO_UPGR_DIR)../_data/versions.yml
+
+###############################################################################
+# Determine whether there's a local yaml installed or use dockerized version.
+# Note in order to install local (faster) yaml: "go get github.com/mikefarah/yaml"
+YAML_CMD:=$(shell which yaml || echo docker run --rm -i $(CALICO_BUILD) yaml)
+
+# Use := so that these V_ variables are computed only once per make run.
+V_CALICOCTL := $(shell cat $(VERSIONS_FILE) | $(YAML_CMD) read - '"v3.0".[0].components.calicoctl.version')
+V_CALICOCTL_V2 :=  $(shell cat $(VERSIONS_FILE) | $(YAML_CMD) read - '"v2.6".[0].components.calicoctl.version')
+
+# Now use ?= to allow the versions derived from versions.yml to be
+# overriden (by the environment).
+CALICOCTL_VER ?= $(V_CALICOCTL)
+CALICOCTL_V2_VER ?= $(V_CALICOCTL_V2)
+
+# Construct the calico/ctl names we'll use to download calicoctl and extract the
+# binaries.
+$(info $(shell printf "%-21s = %-10s\n" "CALICOCTL_VER" $(CALICOCTL_VER)))
+$(info $(shell printf "%-21s = %-10s\n" "CALICOCTL_V2_VER" $(CALICOCTL_V2_VER)))
+CTL_CONTAINER_NAME?=calico/ctl:$(CALICOCTL_VER)
+CTL_CONTAINER_V2_NAME?=calico/ctl:$(CALICOCTL_V2_VER)
+
+###############################################################################
+# calico-upgrade build
+# - Building the calico-upgrade binary in a container
+# - Building the calico-upgrade binary outside a container ("simple-binary")
+# - Building the calico/upgrade image
+###############################################################################
+# Determine which OS / ARCH.
+OS := $(shell uname -s | tr A-Z a-z)
+ARCH := amd64
+
+GIT_VERSION?=$(shell git describe --tags --dirty --always)
+CALICO_UPGRADE_DIR=pkg
+UPGRADE_CONTAINER_NAME?=calico/upgrade
+CALICO_UPGRADE_FILES=$(shell find $(CALICO_UPGRADE_DIR) -name '*.go')
+UPGRADE_CONTAINER_CREATED=$(CALICO_UPGRADE_DIR)/.calico_upgrade.created
+
+CALICO_UPGRADE_BUILD_DATE?=$(shell date -u +'%FT%T%z')
+CALICO_UPGRADE_GIT_REVISION?=$(shell git rev-parse --short HEAD)
+
+LOCAL_USER_ID?=$(shell id -u $$USER)
+
+PACKAGE_NAME?=github.com/projectcalico/calico/calico_upgrade
+
+CALICO_UPGRADE_VERSION ?= $(GIT_VERSION)
+LDFLAGS=-ldflags "-X $(PACKAGE_NAME)/pkg/commands.VERSION=$(CALICO_UPGRADE_VERSION) \
+	-X $(PACKAGE_NAME)/pkg/commands.BUILD_DATE=$(CALICO_UPGRADE_BUILD_DATE) \
+	-X $(PACKAGE_NAME)/pkg/commands.GIT_REVISION=$(CALICO_UPGRADE_GIT_REVISION) -s -w"
+
+LIBCALICOGO_PATH?=none
+
+calico/upgrade: $(UPGRADE_CONTAINER_CREATED)      ## Create the calico/upgrade image
+
+.PHONY: clean-calico-upgrade
+clean-calico-upgrade:
+	docker rmi $(UPGRADE_CONTAINER_NAME):latest || true
+
+# Use this to populate the vendor directory after checking out the repository.
+# To update upstream dependencies, delete the glide.lock file first.
+vendor: glide.yaml
```

Comment:
Make targets shouldn't be directories because directories get freshened by any write inside  them, which means that make doesn't rebuild when it should.  You should use a touch file like we do in Felix/libcalico-go etc.